### ISSUE TITLE
[AWAITING CHECKLIST] [$250] Getting a GBR to "Add a bank account" for the invoice was paid elsewhere

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#53577]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.71-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @shawnborton 
**Slack conversation** (hyperlinked to channel name): [expensify_billpay](https://expensify.slack.com/archives/CSL3XBCCR/p1733320957685139)

## Action Performed:
1. Create a test workspace using your @expensify.com account.
2. Send an invoice from this workspace to another user account (@gmail.com).
3. Log into the receiving account (@gmail.com), and mark the invoice as "Paid elsewhere."
4. Wait for a month and observe
## Expected Result:
No GBR should appear for "Add a bank account" if:
The invoice was marked as "Paid elsewhere."
Bank accounts are already configured in the workspace.
## Actual Result:
The GBR to "Add a bank account" persists in the LHN despite:
The invoice being marked as "Paid elsewhere."
Existing bank accounts already listed under Workspace Settings
Tapping Add a bank account takes to Hmm... it's not here screen

## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [x] iOS: Standalone
- [x] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

https://github.com/user-attachments/assets/2bc3dcb8-c156-4444-97b5-aab3640aa495

https://github.com/user-attachments/assets/b969fe88-fc9f-4ab8-b991-c098af59ccdc

https://github.com/user-attachments/assets/de6df4ba-6ff9-4304-8a10-8175143a7660

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021865094934962379584</li>
        <li>Upwork Job ID: 1865094934962379584</li>
        <li>Last Price Increase: 2024-12-06</li>
<li>Automatic offers: </li>
<ul>
<li>ikevin127 | Contributor | 105255712</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @garrettmknight</details>

**Comment by shawnborton** [comment]
cc @madmax330 @cristipaval @davidcardoza in case we want to keep this internal, see [this Slack thread](https://expensify.slack.com/archives/CSL3XBCCR/p1733320957685139) for more.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

The GBR to "Add a bank account" persists in the LHN despite the invoice being marked as "Paid elsewhere". There are existing bank accounts already listed under Workspace Settings and tapping `Add a bank account` takes to `Hmm... it's not here` screen.

### What is the root cause of that problem?

**There's two main issues here:**

1. The sidebar report GBR is showing because we're not checking whether the invoice report is already paid (settled) here:
- the check is used in `OptionRowLHN`:

https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/components/LHNOptionsList/OptionRowLHN.tsx#L118

and if we follow `requiresAttentionFromCurrentUser` down in `ReportUtils` we arrive to this as the cause:

https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/libs/ReportUtils.ts#L2804-L2808

2. The report action message is showing the `Add a bank account` button because we're not checking whether the invoice report is already paid (settled) here:

https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/pages/home/report/ReportActionItemMessage.tsx#L139-L146

**Secondary reasons for why the GBR will still show in both the sidebar report (which pushes them to the top of the LHN) and in report previews (with 2 or more of invoices) are:**

3. In sidebar report because of the nested reports (with 2 or more of invoices) because we're not checking whether the invoice report is already paid (settled) here:

- the same check used in `OptionRowLHN`:

https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/components/LHNOptionsList/OptionRowLHN.tsx#L118

and if we follow `requiresAttentionFromCurrentUser` down in `ReportUtils` we arrive to this next check as the cause:

https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/libs/ReportUtils.ts#L2810-L2821

where the child reports are being checked instead of the parent (as above in the main issue).

4. In report previews (with 2 or more of invoices) because of the nested reports since we're not checking whether the invoice report is already paid (settled) here:

https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/components/ReportActionItem/ReportPreview.tsx#L349

### What changes do you think we should make in order to solve the problem?

**TLDR:** Add `!ReportUtils.isSettled()` check to all the conditions where this check is missing in order to not show GBR and `Add a bank account` button when invoice was already paid (either through Expensify or elsewhere, it covers both).

**Modify each of the 4 conditions as follows:**

1. https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/libs/ReportUtils.ts#L2804-L2808

becomes:

```ts
if (hasMissingInvoiceBankAccount(optionOrReport.reportID) && !isSettled(optionOrReport.reportID)) {
    return {
        reason: CONST.REQUIRES_ATTENTION_REASONS.HAS_MISSING_INVOICE_BANK_ACCOUNT,
    };
}
```

2. https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/pages/home/report/ReportActionItemMessage.tsx#L139-L146

becomes:

```ts
const shouldShowAddBankAccountButton = action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU && ReportUtils.hasMissingInvoiceBankAccount(reportID) && !ReportUtils.isSettled(reportID);

// ...

{shouldShowAddBankAccountButton && (
    <Button
        style={[styles.mt2, styles.alignSelfStart]}
        success
        text={translate('workspace.invoices.paymentMethods.addBankAccount')}
        onPress={openWorkspaceInvoicesPage}
    />
)}
```

3. https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/libs/ReportUtils.ts#L2810-L2813

becomes:

```ts
if (isInvoiceRoom(optionOrReport)) {
    const reportAction = Object.values(reportActions).find(
        (action) =>
            action.actionName === CONST.REPORT.ACTIONS.TYPE.REPORT_PREVIEW &&
            action.childReportID &&
            hasMissingInvoiceBankAccount(action.childReportID) &&
            !isSettled(action.childReportID),
    );
```

4. https://github.com/Expensify/App/blob/b595f5d24141a403fcc3f44c73d00cd6951dca7d/src/components/ReportActionItem/ReportPreview.tsx#L349

becomes:

```ts
const shouldPromptUserToAddBankAccount =
        (ReportUtils.hasMissingPaymentMethod(userWallet, iouReportID) || ReportUtils.hasMissingInvoiceBankAccount(iouReportID)) && !ReportUtils.isSettled(iouReportID);
```

<details>
<summary>Here's a DIFF of all the changes (for convenience)</summary>

```diff
diff --git a/src/components/ReportActionItem/ReportPreview.tsx b/src/components/ReportActionItem/ReportPreview.tsx
index 41acbf5b4f9..1c3d5205c44 100644
--- a/src/components/ReportActionItem/ReportPreview.tsx
+++ b/src/components/ReportActionItem/ReportPreview.tsx
@@ -346,7 +346,8 @@ function ReportPreview({
 
     const shouldShowSettlementButton = (shouldShowPayButton || shouldShowApproveButton) && !showRTERViolationMessage && !shouldShowBrokenConnectionViolation;
 
-    const shouldPromptUserToAddBankAccount = ReportUtils.hasMissingPaymentMethod(userWallet, iouReportID) || ReportUtils.hasMissingInvoiceBankAccount(iouReportID);
+    const shouldPromptUserToAddBankAccount =
+        (ReportUtils.hasMissingPaymentMethod(userWallet, iouReportID) || ReportUtils.hasMissingInvoiceBankAccount(iouReportID)) && !ReportUtils.isSettled(iouReportID);
     const shouldShowRBR = hasErrors && !iouSettled;
 
     /*
diff --git a/src/libs/ReportUtils.ts b/src/libs/ReportUtils.ts
index ba507ef681c..9e7c41d8da3 100644
--- a/src/libs/ReportUtils.ts
+++ b/src/libs/ReportUtils.ts
@@ -2801,7 +2801,7 @@ function getReasonAndReportActionThatRequiresAttention(
         };
     }
 
-    if (hasMissingInvoiceBankAccount(optionOrReport.reportID)) {
+    if (hasMissingInvoiceBankAccount(optionOrReport.reportID) && !isSettled(optionOrReport.reportID)) {
         return {
             reason: CONST.REQUIRES_ATTENTION_REASONS.HAS_MISSING_INVOICE_BANK_ACCOUNT,
         };
@@ -2809,7 +2809,11 @@ function getReasonAndReportActionThatRequiresAttention(
 
     if (isInvoiceRoom(optionOrReport)) {
         const reportAction = Object.values(reportActions).find(
-            (action) => action.actionName === CONST.REPORT.ACTIONS.TYPE.REPORT_PREVIEW && action.childReportID && hasMissingInvoiceBankAccount(action.childReportID),
+            (action) =>
+                action.actionName === CONST.REPORT.ACTIONS.TYPE.REPORT_PREVIEW &&
+                action.childReportID &&
+                hasMissingInvoiceBankAccount(action.childReportID) &&
+                !isSettled(action.childReportID),
         );
 
         return reportAction
diff --git a/src/pages/home/report/ReportActionItemMessage.tsx b/src/pages/home/report/ReportActionItemMessage.tsx
index da2f3dd151c..24dd0d4558a 100644
--- a/src/pages/home/report/ReportActionItemMessage.tsx
+++ b/src/pages/home/report/ReportActionItemMessage.tsx
@@ -131,12 +131,14 @@ function ReportActionItemMessage({action, displayAsGroup, reportID, style, isHid
         Navigation.navigate(ROUTES.WORKSPACE_INVOICES.getRoute(policyID));
     };
 
+    const shouldShowAddBankAccountButton = action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU && ReportUtils.hasMissingInvoiceBankAccount(reportID) && !ReportUtils.isSettled(reportID);
+
     return (
         <View style={[styles.chatItemMessage, style]}>
             {!isHidden ? (
                 <>
                     {renderReportActionItemFragments(isApprovedOrSubmittedReportAction)}
-                    {action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU && ReportUtils.hasMissingInvoiceBankAccount(reportID) && (
+                    {shouldShowAddBankAccountButton && (
                         <Button
                             style={[styles.mt2, styles.alignSelfStart]}
                             success
```

</details>

### Result screenshots (before / after)

<details>
<summary>MacOS: Chrome</summary>

| Before (parent) | After (parent) |
| ------------- | ------------- |
| <img width="1433" alt="before-2" src="https://github.com/user-attachments/assets/a8b7f81b-78d9-467c-b37d-dce6f09844ee"> | <img width="1433" alt="after-2" src="https://github.com/user-attachments/assets/667bb3f5-4259-4716-b6cc-61324f56545d"> |

| Before (child) | After (child) |
| ------------- | ------------- |
| <img width="1433" alt="before-1" src="https://github.com/user-attachments/assets/6cd6476c-ccef-4738-af9b-dc9a0859112b"> | <img width="1433" alt="after-1" src="https://github.com/user-attachments/assets/90f3b1ff-5575-440d-bfc0-d72e7a7c12b1"> |

</details>

**Comment by allroundexperts** [comment]
Thanks for such a well written proposal @ikevin127.

@ikevin127's [proposal](https://github.com/Expensify/App/issues/53577#issuecomment-2524960552) looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by cristipaval** [comment]
@arosiclair I'm taking over this one as I have context and discussed about it in billpay.

**Comment by ikevin127** [comment]
@cristipaval Hi. Why am I assigned here if you are taking over, should I open the PR or do nothing and will receive partial compensation for providing the RCA / solution ?

**Comment by allroundexperts** [comment]
@ikevin127 I think @cristipaval took over as the internal engineer.

**Comment by ikevin127** [comment]
Thanks, I want to know what's going to happen compensation wise though.

**Comment by garrettmknight** [comment]
Pinged @cristipaval to clarify whether he'll be fixing or just reviewing your fix.

**Comment by cristipaval** [comment]
> @ikevin127 I think @cristipaval took over as the internal engineer.

yes

**Comment by cristipaval** [comment]
Go ahead with the implementation @ikevin127

**Comment by ikevin127** [comment]
Sure, will have a PR open in a few hours.

**Comment by ikevin127** [comment]
PR https://github.com/Expensify/App/pull/53802 is ready for review!

**Comment by ikevin127** [comment]
♻️ **Status update**: PR was open and ready for review for 4 days - pending review from @allroundexperts.

**Comment by garrettmknight** [comment]
PR just hit staging.

**Comment by ikevin127** [comment]
cc @garrettmknight for visibility

**Comment by allroundexperts** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by garrettmknight** [comment]
@allroundexperts request payment when you're ready.

**Comment by JmillsExpensify** [comment]
$250 approved for @allroundexperts
